src/wasm/wasm-debug.cpp

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+
+#include "wasm-debug.h"
+#include "wasm.h"
+
+#ifdef USE_LLVM_DWARF
+#include "thirdparty/llvm/ObjectYAML/DWARFEmitter.h"
+#include "thirdparty/llvm/ObjectYAML/DWARFYAML.h"
+#include "thirdparty/llvm/include/llvm/DebugInfo/DWARFContext.h"
+
+std::error_code dwarf2yaml(llvm::DWARFContext& DCtx, llvm::DWARFYAML::Data& Y);
+#endif
+
+namespace wasm {
+
+namespace Debug {
+
+bool isDWARFSection(Name name) { return name.startsWith(".debug_"); }
+
+bool hasDWARFSections(const Module& wasm) {
+  for (auto& section : wasm.userSections) {
+    if (isDWARFSection(section.name)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+#ifdef USE_LLVM_DWARF
+
+struct BinaryenDWARFInfo {
+  llvm::StringMap<std::unique_ptr<llvm::MemoryBuffer>> sections;
+  std::unique_ptr<llvm::DWARFContext> context;
+
+  BinaryenDWARFInfo(const Module& wasm) {
+    // Get debug sections from the wasm.
+    for (auto& section : wasm.userSections) {
+      if (Name(section.name).startsWith(".debug_")) {
+        // TODO: efficiency
+        sections[section.name.substr(1)] = llvm::MemoryBuffer::getMemBufferCopy(
+          llvm::StringRef(section.data.data(), section.data.size()));
+      }
+    }
+    // Parse debug sections.
+    uint8_t addrSize = 4;
+    context = llvm::DWARFContext::create(sections, addrSize);
+  }
+};
+
+void dumpDWARF(const Module& wasm) {
+  BinaryenDWARFInfo info(wasm);
+  std::cout << "DWARF debug info\n";
+  std::cout << "================\n\n";
+  for (auto& section : wasm.userSections) {
+    if (Name(section.name).startsWith(".debug_")) {
+      std::cout << "Contains section " << section.name << " ("
+                << section.data.size() << " bytes)\n";
+    }
+  }
+  llvm::DIDumpOptions options;
+  options.Verbose = true;
+  info.context->dump(llvm::outs(), options);
+}
+
+//
+// Big picture: We use a DWARFContext to read data, then DWARFYAML support
+// code to write it. That is not the main LLVM Dwarf code used for writing
+// object files, but it avoids us create a "fake" MC layer, and provides a
+// simple way to write out the debug info. Likely the level of info represented
+// in the DWARFYAML::Data object is sufficient for Binaryen's needs, but if not,
+// we may need a different approach.
+//
+// In more detail:
+//
+// 1. Binary sections => DWARFContext:
+//
+//     llvm::DWARFContext::create(sections..)
+//
+// 2. DWARFContext => DWARFYAML::Data
+//
+//     std::error_code dwarf2yaml(DWARFContext &DCtx, DWARFYAML::Data &Y) {
+//
+// 3. DWARFYAML::Data => binary sections
+//
+//     StringMap<std::unique_ptr<MemoryBuffer>>
+//     EmitDebugSections(llvm::DWARFYAML::Data &DI, bool ApplyFixups);
+//
+// For modifying data, like line numberes, we can in theory do that either on
+// the DWARFContext or DWARFYAML::Data; unclear which is best, but modifying
+// the DWARFContext may save us doing fixups in EmitDebugSections.
+//
+
+void updateDWARF(Module& wasm) {
```

Comment:
Currently if this is just for writing, why is it called `updateDWARF`? It seems unlikely that every update to the dwarf info will be in this pass in the future?

Reply:
Oh, I see; it encodes the new DWARF info and "updates" the custom section. I think the name should reflect the fact that the primary thing it does is write/encode the dwarf data. Updating the file is not even actually necessary; e.g. it's possible we'll eventually just write it out to a separate file from Binaryen.